lti_consumer/lti_consumer.py

```diff
@@ -762,8 +774,39 @@ def lti_launch_handler(self, request, suffix=''):  # pylint: disable=unused-argu
         Returns:
             webob.response: HTML LTI launch form
         """
-        lti_consumer = LtiConsumer(self)
-        lti_parameters = lti_consumer.get_signed_lti_parameters()
+        real_user_data = self.extract_real_user_data()
+
+        lti_consumer = self._get_lti1p1_consumer()
+
+        username = None
+        email = None
+        if self.ask_to_send_username and real_user_data['user_username']:
```

Comment:
Since most installations will not be sending username and email, can you move `self.extract_real_user_data()` to be conditional based on `self.ask_to_send_username or self.ask_to_send_email`?

That will avoid looking up the user unless it's needed.

Reply:
Are we sure we want to do that given that `self.extract_real_user_data()` also returns the `user_language`, which is set independently of the values of `self.ask_to_send_username` and `self.ask_to_send_email`. Or do we not care about `user_language` if those values are both false?

Also, with regards to setting values for `username` and `email` from the result of `self.extract_real_user_data()`, if I use an `or` statement as you suggested, should I then subsequently check `self.ask_to_send_username` before setting the username, and `self.ask_to_send_email` before setting the email?

Reply:
Ok, I missed the `user_language`. And the old implementation was also doing the same thing you're doing. So, it's not easy to optimize this right now.